### PR TITLE
feat(finetuning) : add semantic deduplication for synthetic queries generation 

### LINKED ISSUE #20809: [Feature Request]: Semantic Duplication Check for potential duplicated generated in `generate_synthetic_queries_over_documents`
### Feature Description

# Semantic Dedupulication

## Current State: Absence of any Deduplication step

In `llama_index/finetuning/cross_encoders/dataset_gen.py`, `generate_synthetic_queries_over_documents(...)` generates synthetic questions per document chunk using an LLM and aggregates them into a flat list. However, there is currently no deduplication step before returning the final list of questions. Currently, we have (See [full code](https://github.com/run-llama/llama_index/blob/f6b8a6d00d20f222975ac60f3111f0d29d41b462/llama-index-finetuning/llama_index/finetuning/cross_encoders/dataset_gen.py#L35C1-L88C21)):

```py
questions.extend(response_questions)
...
return questions
```
All generated questions are appended and returned as-is, without normalization or deduplication.

This can result in identical questions across multiple chunks, and near-duplicate questions with minor formatting differences. Further down the training pipeline, this can entail redundant training examples in downstream cross-encoder datasets, and inflated dataset size without additional semantic coverage.

This issue becomes more noticeable when `num_questions_per_chunk is large`, and documents contain repetitive structure.

## Proposed Improvement 1: Lightweight deduplication

Before returning questions, do a check for exact matches, preserving order, stripping whitespace, and removing empty strings:

```py
deduped_questions = list(dict.fromkeys(q.strip() for q in questions if q.strip()))
return deduped_questions
```

## Proposed Improvement 2: Semantic Deduplication

**Main goal.** We want near-duplicate detection (paraphrases, trivial rewrites, casing/punctuation changes, “What is X?” vs “Explain X.”)

We may use embedding-based near-duplicate removal -- encode each generated question into a vector embedding and compare it against embeddings of previously accepted questions. If the cosine similarity between a new question and any existing one exceeds a predefined threshold, the new question is treated as a semantic duplicate and discarded. 

To implement, we should first normalize the text (e.g., stripping whitespace, removing numbering or bullet markers, optionally lowercasing), then generate embeddings using a sentence-level model such as OpenAI embeddings or a local sentence-transformers model. 

To ensure scalability, an Approximate Nearest Neighbor (ANN) index (e.g., FAISS or hnswlib) can be used to efficiently retrieve the most similar existing vectors rather than comparing against all prior questions.

**Key tradeoffs to consider.** This semantic deduplication approach requires computing and storing embeddings, as well as maintaining a vector index for similarity search.

### Reason

_No response_

### Value of Feature

Deduplication helps further down the training pipeline as follows:

- Cross-encoder training quality improves with diverse queries
- Prevents redundant positives in ranking datasets
- Reduces training time and memory usage
- Encourages better semantic coverage across chunks

### PR DESCRIPTION
**Pull request by DineshThumma9**
# Description

Fixes : #20809 

## Summary

Adds optional semantic deduplication for synthetic query generation in cross-encoder training.

### Problem
`generate_synthetic_queries_over_documents()` generates duplicate questions across chunks, inflating dataset size  without adding semantic value.

### Solution
Add a deduplication step user can opti-in  made it backward compatibility  user can opt-in or out and uses faiss for faster operations

### Dependencies
- Added: `faiss-cpu>=1.7.0`
- Existing: `sentence-transformers>=2.3.0`

### Usage
```python
questions = generate_synthetic_queries_over_documents(
    documents,
    enable_deduplication=True,  # Opt-in
    similarity_threshold=0.92    # Configurable
    embedding_model_name="all-MiniLM-L6-v2"  #Configurable 
)
```

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [ ] I added new unit tests to cover this change
- [ ] I believe this change is already covered by existing unit tests

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `uv run make format; uv run make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-finetuning/llama_index/finetuning/cross_encoders/dataset_gen.py [hunk 1]
```diff
@@ -50,6 +54,8 @@ def generate_synthetic_queries_over_documents(
         tokenizer=get_tokenizer(),
     )
 
+    from llama_index.llms.openai import OpenAI
```
Comment (AstraBert) [inline llama-index-finetuning/llama_index/finetuning/cross_encoders/dataset_gen.py:57]: Why has the import been moved here?

**Review by AstraBert — COMMENTED** [review]
This is an overkill for deduplication in my opinion. Plus, it is plausible that the fine-tuning package will be deprecated soon, so I would probably keep things as they are

File: llama-index-finetuning/llama_index/finetuning/cross_encoders/dataset_gen.py (hunk 1 above)
Comment (DineshThumma9) [reply llama-index-finetuning/llama_index/finetuning/cross_encoders/dataset_gen.py:57]: My apologies for the misplaced import statement! I moved it temporarily while editing and simply forgot to revert it before pushing.

Regarding the semantic approach being overkill and the potential deprecation of the fine-tuning package:should I replace the FAISS approach with a simple string normalization and exact-match deduplication? is it better to just leave the module exactly as-is?

**Comment by logan-markewich** [comment]
I moved finetuning out of the repo, we aren't supporting this package right now
